# ✅ Chi-Square Feature Selection (Categorical Data)

This method is **specifically designed for categorical features + categorical target**.

If your data looks like:

* encoded categorical features (0/1)
* target = 0/1

👉 **Chi-square is PERFECT**.

---

## 🧠 Core intuition (very simple)

Chi-square checks:

> “Is this feature actually related to the target
> or are they independent?”

If feature and target are **independent** → ❌ useless
If they are **dependent** → ✅ useful

---

## 🧩 Real-life example

Imagine spam dataset:

| has_link | spam |
| -------- | ---- |
| 1        | 1    |
| 1        | 1    |
| 1        | 1    |
| 0        | 0    |
| 0        | 0    |

Strong relationship → very useful feature.

But:

| has_emoji | spam |
| --------- | ---- |
| 0         | 1    |
| 0         | 0    |
| 0         | 1    |
| 0         | 0    |

No relation → useless.

---

## ⚠️ Very important rule (INTERVIEW FAVORITE)

> Chi-square works only with **non-negative values**

That’s why we apply it **after one-hot encoding**.

Binary values (0/1) → perfect.

---

## 🧠 What chi-square actually measures

It compares:

* observed frequency
  vs
* expected frequency

If difference is large → feature matters.

You don’t need math — intuition is enough.

---

## 🔥 Practical example (hands-on)

### Step 1 — Dataset

```python
import pandas as pd

data = {
    "city": ["Delhi", "Mumbai", "Delhi", "Delhi", "Mumbai", "Delhi"],
    "gender": ["Male", "Female", "Male", "Male", "Female", "Male"],
    "device": ["Android", "Android", "iPhone", "Android", "Android", "iPhone"],
    "purchased": [1, 0, 1, 1, 0, 0]
}

df = pd.DataFrame(data)
```

---

### Step 2 — One-hot encode

```python
df_encoded = pd.get_dummies(df, drop_first=False)

X = df_encoded.drop("purchased", axis=1)
y = df_encoded["purchased"]
```

Now everything is numeric (0/1).

---

### Step 3 — Apply Chi-Square

```python
from sklearn.feature_selection import chi2

chi_scores, p_values = chi2(X, y)
```

---

### Step 4 — See results clearly

```python
chi_df = pd.DataFrame({
    "feature": X.columns,
    "chi_score": chi_scores,
    "p_value": p_values
}).sort_values("chi_score", ascending=False)

chi_df
```

---

## 🧠 How to interpret output

### Chi-score:

* higher = stronger relationship

### p-value:

* < 0.05 → important
* > 0.05 → likely useless

Example:

| feature       | chi_score | p_value |
| ------------- | --------- | ------- |
| device_iPhone | 6.21      | 0.01 ✅  |
| city_Mumbai   | 0.02      | 0.88 ❌  |

---

## 🔥 Feature selection using SelectKBest

Instead of manually checking:

```python
from sklearn.feature_selection import SelectKBest, chi2

selector = SelectKBest(score_func=chi2, k=3)

X_selected = selector.fit_transform(X, y)

selected_features = X.columns[selector.get_support()]
selected_features
```

Boom 💥
Top K categorical features selected.

---

## 🧠 Why Chi-Square is powerful

✅ perfect for categorical data
✅ measures real dependency
✅ fast
✅ interpretable
✅ commonly used in NLP
✅ widely asked in interviews

---

## ❌ When NOT to use Chi-Square

* continuous target (regression)
* negative feature values
* raw text
* scaled data with negatives

---

## 🎯 Interview-ready answer

> “Chi-square feature selection evaluates statistical dependency between categorical features and categorical target by comparing observed and expected frequencies.”

🔥 That line = strong impression.

---

## ⚡ Summary (crystal clear)

* categorical features → encode first
* target must be categorical
* values must be non-negative
* higher chi-score = more useful
* p-value confirms significance

---

## 🧠 You now know:

✅ Variance Threshold → removes useless columns
✅ Chi-Square → finds relevant categorical features

This is **real feature selection pipeline**.

---

In [1]:
import pandas as pd

data = {
    "city": ["Delhi", "Mumbai", "Delhi", "Delhi", "Mumbai", "Delhi"],
    "gender": ["Male", "Female", "Male", "Male", "Female", "Male"],
    "device": ["Android", "Android", "iPhone", "Android", "Android", "iPhone"],
    "purchased": [1, 0, 1, 1, 0, 0]
}

df = pd.DataFrame(data)

In [4]:
df_encoded = pd.get_dummies(df,drop_first=False)
X = df_encoded.drop('purchased',axis=1)
y = df['purchased']

In [5]:
from sklearn.feature_selection import chi2
chi_score , p_values = chi2(X,y)

In [6]:
chi_sqr = pd.DataFrame({
    "features":X.columns,
    "chi_score":chi_score,
    "p_value":p_values
})

In [ ]:
chi_sqr
'''Chi-score:
higher = stronger relationship
p-value:
< 0.05 → important
0.05 → likely useless'''

features  chi_score   p_value
0      city_Delhi        1.0  0.317311
1     city_Mumbai        2.0  0.157299
2   gender_Female        2.0  0.157299
3     gender_Male        1.0  0.317311
4  device_Android        0.0  1.000000
5   device_iPhone        0.0  1.000000

## 📊 Your Chi-Square Output

| feature        | chi_score | p_value |
| -------------- | --------- | ------- |
| city_Delhi     | 1.0       | 0.317   |
| city_Mumbai    | 2.0       | 0.157   |
| gender_Female  | 2.0       | 0.157   |
| gender_Male    | 1.0       | 0.317   |
| device_Android | 0.0       | 1.000   |
| device_iPhone  | 0.0       | 1.000   |

---

# 🧠 RULES TO DECIDE (memorize these)

### ✅ Rule 1 — p-value is KING

> If **p-value < 0.05 → keep**
> If **p-value ≥ 0.05 → remove**

Why?

Because p-value tells:

> “Is this relationship statistically significant or just random?”

---

## Apply rule 👇

### ❌ device_Android

* p = 1.0
* means **completely independent**
* target does not care at all

❌ REMOVE

---

### ❌ device_iPhone

* p = 1.0
* zero relationship

❌ REMOVE

---

### ❌ city_Delhi

* p = 0.317 (> 0.05)

❌ REMOVE

---

### ❌ gender_Male

* p = 0.317 (> 0.05)

❌ REMOVE

---

### ⚠️ city_Mumbai

* p = 0.157 (> 0.05)
* not statistically significant

❌ REMOVE (for strict ML)

---

### ⚠️ gender_Female

* p = 0.157 (> 0.05)

❌ REMOVE

---

## 🔥 FINAL RESULT

👉 **Statistically speaking:**

### ❌ KEEP = NONE

### ❌ REMOVE = ALL

And that is **100% correct**.

---

## 😮 Why everything got removed?

Because your dataset is:

* extremely small
* very few samples
* weak relationships
* randomness dominates

📌 Chi-square is **data-hungry**.

With only 5–6 rows, statistics cannot prove dependency.

---

## 🧠 THIS IS A HUGE LEARNING MOMENT

> Feature selection does NOT guarantee features will survive.

Sometimes the correct answer is:

> “No feature is statistically significant.”

That’s not failure — that’s honesty.

---

## 🔥 VERY IMPORTANT REALITY

In real datasets (10k+ rows):

* p-values drop drastically
* true relationships appear
* chi-square becomes powerful

In toy datasets → almost always fails.

---

## 🧠 How professionals decide in practice

### Option 1 — Strict statistical approach

Used in:

* medical ML
* finance
* research

Rule:

```
p_value < 0.05
```

Your result → keep none ✅

---

### Option 2 — Practical ML approach (very common)

We combine **three things**:

1. chi-score ranking
2. domain logic
3. model validation

Example:

Even if p = 0.15
but business logic says device matters
→ we may keep it temporarily.

---

## 🔥 So what do we actually keep?

### Practical rule:

> Keep **top K features by chi-score**,
> even if p-value slightly high.

Example:

```python
chi_df.sort_values("chi_score", ascending=False).head(2)
```

That would give:

* city_Mumbai
* gender_Female

Then model decides later.

---

## 🎯 Final professional rule (IMPORTANT)

| Stage                   | What we do        |
| ----------------------- | ----------------- |
| Early feature selection | Loose rules       |
| Final model             | Strict evaluation |
| Cross-validation        | Final judge       |

👉 Feature selection is **not a one-shot decision**.

---

## 🧠 Interview-perfect explanation

If interviewer asks:

> “What if all p-values are high?”

You say:

> “It usually indicates small dataset or weak statistical power. In such cases, I rank features using chi-square scores and validate them using model performance instead of blindly dropping everything.”

🔥 That answer is VERY strong.

---

## ⚡ Crystal-clear summary

* p-value < 0.05 → statistically significant
* your dataset is too small
* chi-square cannot prove dependency
* removing all features is mathematically correct
* practically, we keep top-ranked features temporarily
* final decision comes from model performance

---

In [10]:
# Feature selection using SelectKBest
from sklearn.feature_selection import SelectKBest,chi2

selector = SelectKBest(score_func=chi2,k=3)
X_selected = selector.fit_transform(X,y)

selected_features = X.columns[selector.get_support()]
selected_features

Index(['city_Mumbai', 'gender_Female', 'gender_Male'], dtype='object')